Users cannot provide custom context

I am trying to use OpenCensus to report Stackdriver metrics from an App Engine app. This does not currently work, since App Engine expects a custom context to be used for all API calls, while the exporter just does `context.Background()` whenever it needs one.

Is there any reason why `stackdriver.NewExporter()` does not accept a custom context?

FYI, this is the specific error I am getting:
```
panic: not an App Engine context

goroutine 14 [running]:
panic(0x1672a60, 0xc0084374e0)
	go/src/runtime/panic.go:491 +0x283
google.golang.org/appengine/internal.fullyQualifiedAppID(0x1cb64c0, 0xc008414080, 0x1d003c0, 0x0)
	google.golang.org/appengine/internal/identity_classic.go:54 +0x95
google.golang.org/appengine/internal.FullyQualifiedAppID(0x1cb64c0, 0xc008414080, 0xc00870d290, 0x1ca4b00)
	google.golang.org/appengine/internal/api_common.go:77 +0x98
google.golang.org/appengine/internal.AppID(0x1cb64c0, 0xc008414080, 0xc00870d290, 0x1)
	google.golang.org/appengine/internal/identity.go:13 +0x35
google.golang.org/appengine.AppID(0x1cb64c0, 0xc008414080, 0xc008669301, 0x33)
	google.golang.org/appengine/identity.go:20 +0x35
golang.org/x/oauth2/google.findDefaultCredentials(0x1cb64c0, 0xc008414080, 0xc0086c5440, 0x4, 0x4, 0x163e6c0, 0x1, 0xc0086c5440)
	golang.org/x/oauth2/google/default.go:65 +0x52f
golang.org/x/oauth2/google.FindDefaultCredentials(0x1cb64c0, 0xc008414080, 0xc0086c5440, 0x4, 0x4, 0x4, 0x4, 0x4)
	golang.org/x/oauth2/google/go19.go:48 +0x53
google.golang.org/api/internal.Creds(0x1cb64c0, 0xc008414080, 0xc008621c20, 0x18, 0x167d580, 0x30)
	google.golang.org/api/internal/creds.go:41 +0x108
google.golang.org/api/transport/grpc.dial(0x1cb64c0, 0xc008414080, 0xf0fd00, 0xc0086c5400, 0x3, 0x4, 0xc008527368, 0xf3ddf4, 0xc0086c5400)
	google.golang.org/api/transport/grpc/dial.go:65 +0x50e
google.golang.org/api/transport/grpc.Dial(0x1cb64c0, 0xc008414080, 0xc0086c5400, 0x3, 0x4, 0xc0086c53c0, 0xc0086c5400, 0x1d)
	google.golang.org/api/transport/grpc/dial.go:37 +0x58
google.golang.org/api/transport.DialGRPC(0x1cb64c0, 0xc008414080, 0xc0086c5400, 0x3, 0x4, 0x1, 0x1, 0x1)
	google.golang.org/api/transport/dial.go:41 +0x53
cloud.google.com/go/monitoring/apiv3.NewMetricClient(0x1cb64c0, 0xc008414080, 0xc0087072f0, 0x1, 0x1, 0xc0087072f0, 0x0, 0x1)
	cloud.google.com/go/monitoring/apiv3/metric_client.go:106 +0xfe
contrib.go.opencensus.io/exporter/stackdriver.newStatsExporter(0xc00870b5a0, 0x20, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	contrib.go.opencensus.io/exporter/stackdriver/stats.go:81 +0x112
contrib.go.opencensus.io/exporter/stackdriver.NewExporter(0xc00870b5a0, 0x20, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	contrib.go.opencensus.io/exporter/stackdriver/stackdriver.go:169 +0x7e
```

--- a/stackdriver.go
+++ b/stackdriver.go
@@ -168,6 +168,15 @@ type Options struct {
 	// default "opencensus_task" label. You should only do this if you know that
 	// the Resource you set uniquely identifies this Go process.
 	DefaultMonitoringLabels *Labels
+
+	// Context allows users to provide a custom context for API calls.
+	//
+	// This context will be used several times: first, to create Stackdriver
+	// trace and metric clients, and then every time a new batch of traces or
+	// stats needs to be uploaded.
+	//
+	// If unset, context.Background() will be used.
+	Context context.Context
 }
 
 // Exporter is a stats.Exporter and trace.Exporter
@@ -180,8 +189,11 @@ type Exporter struct {
 // NewExporter creates a new Exporter that implements both stats.Exporter and
 // trace.Exporter.
 func NewExporter(o Options) (*Exporter, error) {
+	if o.Context == nil {
+		o.Context = context.Background()
+	}
 	if o.ProjectID == "" {
-		creds, err := google.FindDefaultCredentials(context.Background(), traceapi.DefaultAuthScopes()...)
+		creds, err := google.FindDefaultCredentials(o.Context, traceapi.DefaultAuthScopes()...)
 		if err != nil {
 			return nil, fmt.Errorf("stackdriver: %v", err)
 		}

--- a/stats.go
+++ b/stats.go
@@ -78,7 +78,7 @@ func newStatsExporter(o Options, enforceProjectUniqueness bool) (*statsExporter,
 	}
 
 	opts := append(o.MonitoringClientOptions, option.WithUserAgent(userAgent))
-	client, err := monitoring.NewMetricClient(context.Background(), opts...)
+	client, err := monitoring.NewMetricClient(o.Context, opts...)
 	if err != nil {
 		return nil, err
 	}
@@ -150,7 +150,7 @@ func (e *statsExporter) Flush() {
 
 func (e *statsExporter) uploadStats(vds []*view.Data) error {
 	ctx, span := trace.StartSpan(
-		context.Background(),
+		e.o.Context,
 		"contrib.go.opencensus.io/exporter/stackdriver.uploadStats",
 		trace.WithSampler(trace.NeverSample()),
 	)
@@ -163,7 +163,7 @@ func (e *statsExporter) uploadStats(vds []*view.Data) error {
 		}
 	}
 	for _, req := range e.makeReq(vds, maxTimeSeriesPerUpload) {
-		if err := e.c.CreateTimeSeries(ctx, req); err != nil {
+		if err := createTimeSeries(ctx, e.c, req); err != nil {
 			span.SetStatus(trace.Status{Code: 2, Message: err.Error()})
 			// TODO(jbd): Don't fail fast here, batch errors?
 			return err
@@ -464,4 +464,8 @@ var createMetricDescriptor = func(ctx context.Context, c *monitoring.MetricClien
 
 var getMetricDescriptor = func(ctx context.Context, c *monitoring.MetricClient, mdr *monitoringpb.GetMetricDescriptorRequest) (*metric.MetricDescriptor, error) {
 	return c.GetMetricDescriptor(ctx, mdr)
+}
+
+var createTimeSeries = func(ctx context.Context, c *monitoring.MetricClient, ts *monitoringpb.CreateTimeSeriesRequest) error {
+	return c.CreateTimeSeries(ctx, ts)
 }

--- a/stats_test.go
+++ b/stats_test.go
@@ -16,6 +16,7 @@ package stackdriver
 
 import (
 	"context"
+	"fmt"
 	"reflect"
 	"testing"
 	"time"
@@ -1011,6 +1012,62 @@ func TestExporter_makeReq_withCustomMonitoredResource(t *testing.T) {
 				t.Errorf("%v: Exporter.makeReq() = \n      %v\nwant: %v", tt.name, resps, tt.want)
 			}
 		})
+	}
+}
+
+func TestExporter_customContext(t *testing.T) {
+	oldCreateMetricDescriptor := createMetricDescriptor
+	oldCreateTimeSeries := createTimeSeries
+
+	defer func() {
+		createMetricDescriptor = oldCreateMetricDescriptor
+		createTimeSeries = oldCreateTimeSeries
+	}()
+
+	var timedOut = 0
+	createMetricDescriptor = func(ctx context.Context, c *monitoring.MetricClient, mdr *monitoringpb.CreateMetricDescriptorRequest) (*metric.MetricDescriptor, error) {
+		select {
+		case <-time.After(15 * time.Millisecond):
+			fmt.Println("createMetricDescriptor did not time out")
+		case <-ctx.Done():
+			timedOut++
+		}
+		return &metric.MetricDescriptor{}, nil
+	}
+	createTimeSeries = func(ctx context.Context, c *monitoring.MetricClient, ts *monitoringpb.CreateTimeSeriesRequest) error {
+		select {
+		case <-time.After(15 * time.Millisecond):
+			fmt.Println("createTimeSeries did not time out")
+		case <-ctx.Done():
+			timedOut++
+		}
+		return nil
+	}
+
+	v := &view.View{
+		Name:        "test_view_count",
+		Description: "view_description",
+		Measure:     stats.Float64("test-measure/TestExporter_createMeasure", "measure desc", stats.UnitMilliseconds),
+		Aggregation: view.Count(),
+	}
+
+	data := &view.CountData{Value: 0}
+	vd := newTestViewData(v, time.Now(), time.Now(), data, data)
+
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Millisecond)
+	defer cancel()
+	e := &statsExporter{
+		createdViews: make(map[string]*metricpb.MetricDescriptor),
+		o:            Options{ProjectID: "test_project", Context: ctx},
+	}
+	if err := e.uploadStats([]*view.Data{vd}); err != nil {
+		t.Errorf("Exporter.uploadStats() error = %v", err)
+	}
+	if ctx.Err() != context.DeadlineExceeded {
+		t.Errorf("expected context to time out; got %v", ctx.Err())
+	}
+	if timedOut != 2 {
+		t.Errorf("expected two functions to time out; got %d", timedOut)
 	}
 }
 

--- a/trace.go
+++ b/trace.go
@@ -15,7 +15,6 @@
 package stackdriver
 
 import (
-	"context"
 	"fmt"
 	"log"
 	"sync"
@@ -43,7 +42,7 @@ type traceExporter struct {
 var _ trace.Exporter = (*traceExporter)(nil)
 
 func newTraceExporter(o Options) (*traceExporter, error) {
-	client, err := tracingclient.NewClient(context.Background(), o.TraceClientOptions...)
+	client, err := tracingclient.NewClient(o.Context, o.TraceClientOptions...)
 	if err != nil {
 		return nil, fmt.Errorf("stackdriver: couldn't initialize trace client: %v", err)
 	}
@@ -118,7 +117,7 @@ func (e *traceExporter) uploadSpans(spans []*trace.SpanData) {
 	}
 	// Create a never-sampled span to prevent traces associated with exporter.
 	ctx, span := trace.StartSpan( // TODO: add timeouts
-		context.Background(),
+		e.o.Context,
 		"contrib.go.opencensus.io/exporter/stackdriver.uploadSpans",
 		trace.WithSampler(trace.NeverSample()),
 	)

